### PR TITLE
[data] Add iceberg datasource

### PR DESCRIPTION
**Pull request by dev-goyal**
## Why are these changes needed?

This PR adds the capability to load an Iceberg table into a Ray Dataset. Compared to the PyIceberg functionality, which can only materialize the entire Iceberg table into a single `pyarrow` table first, which is then converted to a Ray dataset, this PR allows a streaming implementation, where the Iceberg table can be distributed into a Ray Dataset.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/datasource/__init__.py [hunk 1]
```diff
@@ -1,3 +1,5 @@
+from iceberg_datasource import IcebergDatasource
```
Comment (dev-goyal) [inline python/ray/data/datasource/__init__.py:1]: Will fix this - for now I was unable to do a pip editable install on my machine, so just had this for testing

File: python/ray/data/datasource/iceberg_datasource.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        self._selected_fields = selected_fields
+        self._snapshot_id = snapshot_id
+
+    def _get_catalog(self) -> "Catalog":
+        from pyiceberg import catalog
+
+        if self._catalog_type == "glue":
+            # Glue catalogs have no name, and no conf options are needed as long as AWS
+            # credentials are set up
+            return catalog.load_glue(name="", conf=self._catalog_kwargs)
```
Comment (Fokko) [inline python/ray/data/datasource/iceberg_datasource.py:75]: I would highly recommend using the `load_catalog` method instead since it offers much higher flexibility.

```suggestion
            if "type" not in self._catalog_kwargs:
                self._catalog_kwargs["type"] = self._catalog_type
            return catalog.load_catalog(name="default", **self._catalog_kwargs)
```

Also, setting a name allows the user to pass in configuration [through environment variables, or a `.pyiceberg.yaml` config](https://py.iceberg.apache.org/configuration/).

File: python/ray/data/datasource/iceberg_datasource.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+        def _get_read_task(
+            tasks: Iterable["FileScanTask"],
+            table_identifier: str,
+            schema: "Schema",
+            catalog_type: str = "glue",
+        ) -> Iterable[block.Block]:
+            # Closure so we can pass this callable as an argument to ReadTask
+
+            # Both the catalog and tbl attributes cannot be pickled, which means they
```
Comment (Fokko) [inline python/ray/data/datasource/iceberg_datasource.py:131]: I think we should be able to do this:

- https://github.com/apache/iceberg-python/issues/513
- https://github.com/apache/iceberg-python/issues/514

File: python/ray/data/datasource/iceberg_datasource.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+                # parquet file, as there might be delete files, etc. associated, so we
+                # must use the PyIceberg API for the projection.
+                yield pyi_pa_io.project_table(
+                    tasks=[task],
+                    table=tbl,
+                    row_filter=self._row_filter,
+                    projected_schema=schema,
+                    case_sensitive=self._scan_kwargs.get("case_sensitive", True),
+                    limit=self._scan_kwargs.get("limit"),
+                )
```
Comment (Fokko) [inline python/ray/data/datasource/iceberg_datasource.py:149]: It would be great to eventually push the deletes and projection to the engine itself. Would it be possible to pass in the group of tasks?

```suggestion
            # Use the PyIceberg API to read only a single task (specifically, a
            # FileScanTask) - note that this is not as simple as reading a single
            # parquet file, as there might be delete files, etc. associated, so we
            # must use the PyIceberg API for the projection.
            yield pyi_pa_io.project_table(
                tasks=tasks,
                table=tbl,
                row_filter=self._row_filter,
                projected_schema=schema,
                case_sensitive=self._scan_kwargs.get("case_sensitive", True),
                limit=self._scan_kwargs.get("limit"),
            )
```

This will be much more efficient

**Review by Fokko — COMMENTED** [review]
Looking good @deanwampler! Excited to see this!

File: python/ray/data/datasource/iceberg_datasource.py (hunk 3 above)
Comment (rdblue) [reply python/ray/data/datasource/iceberg_datasource.py:131]: @Fokko, do we want to be able to pickle a table and a catalog? There are definitely reasons to want to ship a table to another process, but we've found that this can also enable people to do problematic things accidentally, like planning scans in parallel. Not being able to pickle ensures that planning happens once and tasks (which should be picklable) are sent instead.

Comment (Fokko) [reply python/ray/data/datasource/iceberg_datasource.py:131]: That's a good question. I think `Table` makes sense, but I can see that the `Catalog` is less obvious:

- **Table**: You can already serialize the FileIO today. I think being able to serialize the Table makes sense since this is more-or-less just serializing the TableMetadata. We can reconstruct it on the instance itself.
- **Catalog**: This might also make sense if we want to refresh credentials and such, but looking at the code here, we could also avoid pickling the catalog.

**Comment by dev-goyal** [comment]
Thanks @Fokko , great comments! Glad this PR has some traction - I'm going to look into your review soon!

File: python/ray/data/datasource/iceberg_datasource.py (hunk 4 above)
Comment (dev-goyal) [reply python/ray/data/datasource/iceberg_datasource.py:149]: I see what you mean here, but I was mostly worried about (1) reducing parallelism too much and (2) pushing too much computation down to the pyiceberg engine, i.e., what if all of `[tasks]` is too big to be held within a single read task?

**Comment by dev-goyal** [comment]
> Thanks @Fokko , great comments! Glad this PR has some traction - I'm going to look into your review soon!

@Fokko incorporated all your comments - let me know what you think!

**Comment by zhe-thoughts** [comment]
Thanks for the contribution @dev-goyal !

Quick comment: please remember to add a test

**Comment by dev-goyal** [comment]
> Thanks for the contribution @dev-goyal !
> 
> Quick comment: please remember to add a test

Hi @zhe-thoughts - added a bunch of tests - what do you think?

**Comment by anyscalesam** [comment]
@c21 can you review the latest test adds from @dev-goyal so we can see if we can merge this?

File: doc/source/data/loading-data.rst [hunk 5]
```diff
@@ -919,6 +919,31 @@ run against the collection.
...
+    from pyiceberg.expressions import BoolExpression, EqualTo
+
+    # Read an iceberg table
+    ds = ray.data.read_iceberg(
+        table_identifier="db_name.table_name",
+        row_filter=BooleanExpression(EqualTo("column_name", "literal_value")),
+        catalog_kwargs={"name": "default", "type": "glue"},
+    )
+
+Writes are currently not supported, but might be soon as PyIceberg upgrades to support partitioned writes, etc.
```
Comment (aslonnie) [inline doc/source/data/loading-data.rst:945]: maybe remove `but might be soon as PyIceberg upgrades to support partitioned writes, etc.`

File: doc/source/data/loading-data.rst [hunk 6]
```diff
@@ -919,6 +919,31 @@ run against the collection.
         collection="my_collection",
     )
 
+Reading Iceberg
+~~~~~~~~~~~~~~~
+
+To read data from Iceberg, call :func:`~ray.data.read_iceberg` and specify the catalog ID, namespace (database), and table.
+To read from Iceberg, install the `Python Client for Iceberg <https://py.iceberg.apache.org/>`_.
```
Comment (aslonnie) [inline doc/source/data/loading-data.rst:926]: "To read data from Iceberg" and "To read from Iceberg" feels repeating. could you rephrase?

**Comment by dev-goyal** [comment]
Apologies, closed this by mistake in an effort to rebase, new PR here

https://github.com/ray-project/ray/pull/45007
